### ISSUE TITLE
Angular Components (CDK and Material) v12 documentation example exceptions

### ISSUE THREAD

**Issue by Splaktar** [stackblitz/core#1405]
## Hello StackBlitz team!

We're trying to debug some issues with the Angular Material examples in StackBlitz. We had some similar problems on our own docs site and we solved them today by updating TypeScript to 4.2.3 since Angular v12 now requires TypeScript 4.2.x (and no longer supports TypeScript 4.0.x or 4.1.x).

### In researching this, I found that

- https://stackblitz.com/fork/angular-ivy uses TypeScript ~4.0.2, which makes sense for [Angular v10](https://github.com/angular/angular-cli/blob/v10.2.3/packages/schematics/angular/utility/latest-versions.ts) (but not for v11 which [needs TypeScript ~4.1.x](https://github.com/angular/angular-cli/blob/v11.2.5/packages/schematics/angular/utility/latest-versions.ts))
- I found https://github.com/stackblitz/angular-cli-template, but it seems to have been abandoned at Angular v5

### What we're having trouble with

1. https://next.material.angular.io/components/autocomplete/examples

```ts
preview-c9f765cddfc622204e06b.js:1 ERROR TypeError: Cannot read property 'autoActiveFirstOption' of undefined
    at MatAutocomplete._MatAutocompleteBase [as constructor] (autocomplete.ts:215)
    at new MatAutocomplete (autocomplete.ts:280)
    at createClass (provider.ts:273)
    at createDirectiveInstance (provider.ts:142)
    at createViewNodes (view.ts:314)
    at callViewAction (view.ts:649)
    at execComponentViewsAction (view.ts:572)
    at createViewNodes (view.ts:344)
    at createRootView (view.ts:216)
    at callWithDebugContext (services.ts:641)
```
What this leads to is basically that DI is not working as expected:
```ts
@Inject(MAT_AUTOCOMPLETE_DEFAULT_OPTIONS) defaults: MatAutocompleteDefaultOptions,
```
[defaults](https://github.com/angular/components/blob/71f83867ddf4615a331494ce494fafa51c027dfe/src/material/autocomplete/autocomplete.ts#L215) ends up as `undefined` since the above injection doesn't work.

2. https://next.material.angular.io/components/select/examples

```ts
@Inject(MAT_SELECT_SCROLL_STRATEGY) scrollStrategyFactory: any,
```
[this._scrollStrategyFactory](https://github.com/angular/components/blob/7a972fbe38abc2401514dd2c24424faec645d2bb/src/material/select/select.ts#L489) ends up as `undefined` since the above injection doesn't work.

### What we've tried

1. `"enableIvy": true,` in our template's `tsconfig.json`. This resulted in a TypeScript exception and a ton of CORS errors. I'm not sure if this is related to [this SO question about `ngcc` in StackBlitz](https://stackoverflow.com/questions/63386983/stackblitz-ngcc-failed-to-run-on-my-npm-library-with-ivy-enabled), but we don't get "ngcc failed to run". Just this:

    ```ts
    Uncaught (in promise) TypeError: Cannot read property 'native' of undefined
    at getNodeSystem (typescript.js:7202)
    at eval (typescript.js:7753)
    at eval (typescript.js:7760)
    at Object.eval (typescript.js:7770)
    at eval (typescript.js:153198)
    at eval (typescript.js:153199)
    at eval (<anonymous>)
    at Qt (webcontainer.3a09fea6d6af8fa50aecb2140d2b6d2039950e89.js:15)
    at webcontainer.3a09fea6d6af8fa50aecb2140d2b6d2039950e89.js:15
    at U (webcontainer.3a09fea6d6af8fa50aecb2140d2b6d2039950e89.js:15)
    Access to fetch at 'https://l.staticblitz.com/ngcc/v3/12.0.0-next.5/@angular/cdk@12.0.0-next.3' from origin 'https://stackblitz.com' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource. If an opaque response serves your needs, set the request's mode to 'no-cors' to fetch the resource with CORS disabled.
    ```
2. Update `"typescript": "~4.2.3"` in our template's `devDependencies`, this has no effect. This and the next bullet point seem related to https://github.com/stackblitz/core/issues/483.
3. Update `"typescript": "~4.2.3"` in our template's `dependencies`, this has no effect.
4. These same examples work on https://material.angular.io and in our `dev-app` in the `angular/components` repo. There are only exceptions on StackBlitz.

### What's next

Please let us know if
- we are doing something wrong or you would like us to try something to gather more info
- this is a known issue that is being worked on
- you are fairly confident that this isn't a StackBlitz issue

**Comment by EricSimons** [comment]
@Splaktar are these versions of the CDK and Material being distributed for the new Linker shipping in Angular v12? i.e. would they work with versions of Angular core and CLI that are lower than 12? (If not, then this might be the main problem, as linker support is WIP on our side)

**Comment by Splaktar** [comment]
I am not aware of any hard dependency on the Angular Linker at this point, but @petebacondarwin may know more. I would assume that if that was the issue, then none of our examples would work. However only our autocomplete and select are having issues.

**Comment by petebacondarwin** [comment]
All the versions of Angular packages (core framework and components) are still being distributed as ViewEngine builds, so there is no "linker" aspect to this. If ivy is enabled in the application project then ngcc will need to run on them.

**Comment by alexzuza** [comment]
@petebacondarwin 
I think ViewEngine doesn't work here because new TypeScript version changed transpiled to ES5 code for extended classes.

Angular compiler can handle 
```
_super.apply(this, arguments) || this;

_super.apply(this, __spread(arguments)) || this;
```
but not

```
_super.apply(this, __spreadArray([], __read(arguments)))
```
This is what we have in latest `material-autocomplete.umd.js`

https://github.com/angular/material-builds/blob/185bb52ae51b1d85cd2e28a090b0e1ab383bfc9e/bundles/material-autocomplete.umd.js#L497-L499

Related code can be found here:
https://github.com/angular/angular/blob/a6971ba89adc253bfa4260036ee4a1e0bd76159f/packages/core/src/reflection/reflection_capabilities.ts#L26-L45

**Comment by petebacondarwin** [comment]
We landed a fix for that in `master` recently: https://github.com/angular/angular/pull/41201

**Comment by Splaktar** [comment]
It looks like [that fix](https://github.com/angular/angular/commit/7b1214eca2dd2f09e723a46bed857fcb7d40bc0b) landed in Angular `12.0.0-next.5` which we're using in the above examples.

**Comment by alexzuza** [comment]
@Splaktar Yep, that fix was landed in 12.0.0-next.5. But it is for ngcc. I'm talking about **ViewEngine** which is default for Angular Material demos.

ViewEngine reflector can't read correctly metadata from umd es5 bundle. This RegExp needs to be adjusted https://github.com/angular/angular/blob/a6971ba89adc253bfa4260036ee4a1e0bd76159f/packages/core/src/reflection/reflection_capabilities.ts#L44-L45

**Comment by JoostK** [comment]
@alexzuza excellent find! Yes, indeed, the support in angular/angular#41201 is limited to partial evaluation of the new spread helper. We do indeed also have to update the reflector for JIT compiles, and additionally ngcc needs a fix for ES5 synthesized constructors:

https://github.com/angular/angular/blob/0accd1e68dc1c5b386a4531bc9c99f18b930ddfd/packages/compiler-cli/ngcc/src/host/esm5_host.ts#L358-L392

I was initially thinking that latest APF no longer contains ES5 bundles, but as you pointed out this happens in UMD bundles as they are ES5.

**Comment by JoostK** [comment]
The problem with UMD bundles should be resolved in Angular FW `12.0.0-next.6`.

**Comment by Splaktar** [comment]
Confirmed that our examples now run without issues now since they automatically install Angular `12.0.0-next.7`.
